### PR TITLE
[Reviewer: Ellie] Parses the config option snmp_notification_type to the alarm trap sender

### PR DESCRIPTION
**Pull request by AME2**

### CODE REVIEW

**Comment by AME2** [comment]
I should have mentioned the only testing of this I've done is coverage and live testing.

**Comment by eleanor-merry** [comment]
Passing back to you while you make updates

File: clearwater-snmp-alarm-agent.root/usr/share/clearwater/bin/clearwater-snmp-alarm-agent-wrapper [hunk 1]
```diff
@@ -47,7 +48,7 @@ get_daemon_args()
   # Don't try and load MIBS at startup - we don't need them and this just causes lots of
   # "missing MIB" error logs.
   export MIBS=""
-  DAEMON_ARGS="--snmp-ips $snmp_ip --local_ip $local_ip --community $snmp_community --log-dir $log_directory --log-level $log_level"
+  DAEMON_ARGS="--snmp-ips $snmp_ip --local_ip $local_ip --community $snmp_community --snmp_notification_type $snmp_notification_type --log-dir $log_directory --log-level $log_level"
```
Comment (eleanor-merry) [inline clearwater-snmp-alarm-agent.root/usr/share/clearwater/bin/clearwater-snmp-alarm-agent-wrapper]: We normally use - for config options, not _

File: include/alarm_trap_sender.hpp [hunk 2]
```diff
@@ -43,15 +43,23 @@
...
 // Class providing methods for generating alarmActiveState and
 // alarmClearState inform notifications.
 class AlarmTrapSender
 {
 public:
-  void initialise(AlarmScheduler* alarm_scheduler)
-    { _alarm_scheduler = alarm_scheduler; }
+  void initialise(AlarmScheduler* alarm_scheduler,
+                  std::vector<NotificationType> snmp_notifications)
+    { _alarm_scheduler = alarm_scheduler;
```
Comment (eleanor-merry) [inline include/alarm_trap_sender.hpp]: Not a fan of this alignment

File: src/alarms_agent.cpp [hunk 3]
```diff
@@ -62,6 +62,7 @@ void agent_terminate_handler(int sig)
 enum OptionTypes
 {
   OPT_COMMUNITY=256+1,
+  OPT_NOTIFICATION_TYPE,
```
Comment (eleanor-merry) [inline src/alarms_agent.cpp]: Why not OPT_SNMP_NOTIFICATION_TYPE?

File: src/alarms_agent.cpp [hunk 4]
```diff
@@ -83,6 +85,7 @@ static void usage(void)
          "\n"
          " --snmp-ips <ip>,<ip>       Send SNMP notifications to the specified IPs\n"
          " --community <name>         Include the given community string on notifications\n"
+         " --snmp_notification_type   Sends SNMP noticiations with the specified format\n"
```
Comment (eleanor-merry) [inline src/alarms_agent.cpp]: *notifications

File: src/alarms_agent.cpp [hunk 5]
```diff
@@ -107,6 +111,37 @@ int main (int argc, char **argv)
...
         break;
+      case OPT_NOTIFICATION_TYPE:
+        {
+          std::vector<std::string> notification_types;
+          Utils::split_string(std::string(optarg), ',', notification_types);
+          for (std::vector<std::string>::iterator it = notification_types.begin();
+               it != notification_types.end();
+               ++it)
+          {
+            if (*it == "rfc3877" && std::find(snmp_notifications.begin(),
```
Comment (eleanor-merry) [inline src/alarms_agent.cpp]: As discussed, this is easier if you use a set rather than a vector here

File: src/alarms_agent.cpp [hunk 6]
```diff
@@ -107,6 +111,37 @@ int main (int argc, char **argv)
...
+            }
+            else if (*it == "enterprise" && std::find(snmp_notifications.begin(), 
+                                              snmp_notifications.end(),
+                                              NotificationType::ENTERPRISE) == snmp_notifications.end())
+            {
+              snmp_notifications.push_back(NotificationType::ENTERPRISE);
+            }
+            else
+            {
+              puts("Invalid config option used for snmp notification type");
```
Comment (eleanor-merry) [inline src/alarms_agent.cpp]: This would be more useful if it included what the invalid config was

File: src/alarms_agent.cpp [hunk 7]
```diff
@@ -107,6 +111,37 @@ int main (int argc, char **argv)
...
+            }
+          }
+          if (snmp_notifications.empty())
```
Comment (eleanor-merry) [inline src/alarms_agent.cpp]: Add a blank line above the if

File: clearwater-snmp-alarm-agent.root/usr/share/clearwater/bin/clearwater-snmp-alarm-agent-wrapper (hunk 1 above)
Comment (eleanor-merry) [inline clearwater-snmp-alarm-agent.root/usr/share/clearwater/bin/clearwater-snmp-alarm-agent-wrapper]: Have you tested what happens if a user sets snmp_notification_type= in the config file?

File: src/alarms_agent.cpp (hunk 7 above)
Comment (eleanor-merry) [inline src/alarms_agent.cpp]: Actually, this is in the wrong place as it means snmp_notifications is only populated if a user passes in something under snmp_notification_type. I assume you didn't see this in your testing as you've set the default in the wrapper script as well. I think you should take the default out of the wrapper script and move this check until after all the parameters have been parsed.

**Comment by eleanor-merry** [comment]
Generally looks good; there's one issue you need to fix (the check for notification types being empty), and I've made a number of comments for how to improve the code.

However, I don't think this will compile when you run the tests (you've changed the AlarmScheduler constructor but not changed any test code) - can you make any UT changes necessary as well.

File: clearwater-snmp-alarm-agent.root/usr/share/clearwater/bin/clearwater-snmp-alarm-agent-wrapper [hunk 8]
```diff
@@ -38,6 +38,7 @@ get_daemon_args()
 {
   snmp_ip=0.0.0.0
   snmp_community=clearwater
+  snmp_notification_type=rfc3877
```
Comment (rkd-msw) [inline clearwater-snmp-alarm-agent.root/usr/share/clearwater/bin/clearwater-snmp-alarm-agent-wrapper]: The spec says the config option is called "types" - is this right?

Comment (AME2) [inline clearwater-snmp-alarm-agent.root/usr/share/clearwater/bin/clearwater-snmp-alarm-agent-wrapper]: Yes my mistake, I'll change this

**Pull request merged**